logutil/pretty_handler.go

```diff
@@ -0,0 +1,330 @@
+			if s.colorFunc != nil {
+				str = s.colorFunc(str)
+			}
+			b.WriteString(str)
+		} else {
+			// If the level was modified by ReplaceAttrs then it is the caller's
+			// responsibility to handle colouring.
+			b.WriteString(stringify(a.Value))
+		}
+	} else if a.Key == slog.SourceKey {
+		b.WriteString(h.c.Magenta(stringify(a.Value)))
+	} else if a.Key == slog.MessageKey {
+		fmt.Fprintf(b, "%-44s", stringify(a.Value))
+	} else {
+		// Handle remaining attrs.
+		h.appendString(b, s.groupPrefix+a.Key, s.colorFunc)
+		b.WriteByte('=')
+		h.appendString(b, stringify(a.Value), nil)
+	}
+	b.WriteByte(' ')
+}
+
+type state struct {
+	groupPrefix string
+	groups      []string
+	colorFunc   func(string) string
+}
+
+func (h *PrettyHandler) appendString(b *bytes.Buffer, s string, colorFunc func(string) string) {
+	if h.needsQuoting(s) {
+		s = strconv.Quote(s)
+	}
+	if colorFunc != nil {
+		s = colorFunc(s)
+	}
+	b.WriteString(s)
+}
+
+func (h *PrettyHandler) needsQuoting(s string) bool {
+	if h.opts.ForceQuote || s == "" {
+		return true
+	}
+	for _, c := range s {
+		// Needs to be quoted if it's not alphanumeric and not one of the special chars below.
+		if !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') ||
+			c == '-' || c == '.' || c == '_' || c == '/' || c == '@' || c == '^' || c == '+' || c == ':') {
+			return true
+		}
+	}
+	return false
+}
+
+func stringify(v slog.Value) string {
+	switch v.Kind() {
+	case slog.KindBool:
+		return strconv.FormatBool(v.Bool())
+	case slog.KindInt64:
+		return strconv.FormatInt(v.Int64(), 10)
+	case slog.KindUint64:
+		return strconv.FormatUint(v.Uint64(), 10)
+	case slog.KindFloat64:
+		return strconv.FormatFloat(v.Float64(), 'g', -1, 64)
+	case slog.KindString:
+		return v.String()
+	case slog.KindDuration:
+		return v.Duration().String()
+	case slog.KindTime:
+		return v.Time().Format(time.RFC3339)
+	case slog.KindAny:
+		vv := v.Any()
+		if tm, ok := vv.(encoding.TextMarshaler); ok {
+			data, err := tm.MarshalText()
+			if err != nil {
+				return fmt.Sprintf("!ERROR:%v", err)
+			}
+			return string(data)
+		}
+		// Handle byte slices specially and try and print it nicely.
+		if bs, ok := vv.([]byte); ok {
+			return string(bs)
+		}
+		if t := reflect.TypeOf(vv); t != nil {
+			// Handle the underlying type being []byte.
+			if t.Kind() == reflect.Slice && t.Elem().Kind() == reflect.Uint8 {
+				return string(reflect.ValueOf(vv).Bytes())
+			}
+			// Check to see if a we have a function or channel.
+			// These can't be printed properly and fmt.Sprint will generate an
+			// ugly pointer representation, ex: %!s(func()=0x10bf380)
+			switch t.Kind() {
+			case reflect.Chan:
+				name := t.Elem().Name()
+				if name == "" {
+					name = "unknown"
+				}
+				return "chan " + name
+			case reflect.Func:
+				return "func()"
+			}
+		}
+
+		return fmt.Sprintf("%+v", v)
+	default:
+		panic(fmt.Errorf("impossible: invalid slog.Value kind: %s", v.Kind()))
```

Comment:
Can we actually get to here? I feel like a logger shouldn't panic if it can't format something

Reply:
It's an impossible case just for the sake of completeness, the `case slog.KindAny:` above is the real catch-all case

Reply:
It has `impossible:` in the message 🙂